### PR TITLE
add subword unit option.

### PR DESCRIPTION
**Pull request by jnishi**
Add subword unit option in Librispeech recipe. This PR is still work in progress. I am going to add features in this PR below.

- [x] support installation of sentencepiece.
- [x] support training with sentencepiece.
- [x] support decoding with sentencepiece 
- [x] support evaluation
- [x] support RNNLM
- [x] fix egs/wsj/asr1/local/run_bpe.sh to use sentencepiece (or remove it)

### CODE REVIEW

**Comment by jnishi** [comment]
I supported sentencepiece tokenizer in librispeech recipe. But the accuracy of the neural network is extremely worse than character based neural network. The accuracy after training is like below.

```
    {
        "main/loss": 219.3127505493164,
        "main/loss_ctc": 238.20053699493408,
        "iteration": 62300,
        "eps": 1e-10,
        "main/loss_att": 200.42496566772462,
        "elapsed_time": 130545.5796751976,
        "epoch": 14,
        "main/acc": 0.3010910286248617
    }
```

I think this is because learning rate must be larger in short sequence than in long sequence. I'm going to check to train with different parameters.

**Comment by jnishi** [comment]
@sw005320 
The accuracy of the neural network seems to be at most 30%.
Do you have results of recognition with WSJ with subword-nmt?

**Comment by jnishi** [comment]
Thanks. I'm going to try 200 pieces.

**Comment by jnishi** [comment]
Added RNNLM.
Using 200 pieces is still worse, but using 60 pieces seems to be good. 

@sw005320 Would you review this?

File: egs/librispeech/asr1/local/run_spm.sh [hunk 1]
```diff
@@ -231,7 +262,7 @@ if [ ${stage} -le 5 ]; then
 
     for rtask in ${recog_set}; do
     (
-        decode_dir=decode_${rtask}_beam${beam_size}_e${recog_model}_p${penalty}_len${minlenratio}-${maxlenratio}_ctcw${ctc_weight}_rnnlm${lm_weight}
+        decode_dir=decode_${rtask}_beam${beam_size}_e${recog_model}_p${penalty}_len${minlenratio}-${maxlenratio}_ctcw${ctc_weight}
```
Comment (sw005320) [inline egs/librispeech/asr1/local/run_spm.sh]: I think it better to add `_rnnlm${lm_weight}`

File: egs/librispeech/asr1/local/run_spm.sh [hunk 2]
```diff
@@ -241,20 +272,24 @@ if [ ${stage} -le 5 ]; then
         # feature extraction
```
Comment (sw005320) [inline egs/librispeech/asr1/local/run_spm.sh]: This feature extraction part in the recognition was changed. Please check the diff of `run.sh` in https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/run.sh
Now basic feature dump was moved to the stage 1, and you don't need to use `feats` now.

File: egs/librispeech/asr1/local/run_spm.sh [hunk 3]
```diff
@@ -57,37 +58,62 @@ minlenratio=0.0
...
+# if you're not on the CLSP grid.
+datadir=/export/a15/vpanayotov/data
 
-# bpe
-nbpe=500
+# base url for downloads.
+data_url=www.openslr.org/resources/12
+
+# bpemode (unigram or bpe)
+nbpe=1000
```
Comment (sw005320) [inline egs/librispeech/asr1/local/run_spm.sh]: maybe better to use your current best setup.

File: egs/librispeech/asr1/local/run_spm.sh [hunk 4]
```diff
@@ -98,81 +124,97 @@ if [ ${stage} -le 1 ]; then
     echo "stage 1: Feature Generation"
     fbankdir=fbank
     # Generate the fbank features; by default 80-dimensional fbanks with pitch on each frame
-    for x in train_si284 test_dev93 test_eval92; do
-        steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 10 data/${x} exp/make_fbank/${x} ${fbankdir}
+    for x in dev_clean test_clean dev_other test_other train_clean_100 train_clean_360 train_other_500; do
+        steps/make_fbank_pitch.sh --cmd "$train_cmd" --nj 32 data/${x} exp/make_fbank/${x} ${fbankdir}
     done
 
+    utils/combine_data.sh data/${train_set} data/train_clean_100 data/train_clean_360 data/train_other_500
```
Comment (sw005320) [inline egs/librispeech/asr1/local/run_spm.sh]: Can you check https://github.com/espnet/espnet/blob/master/egs/librispeech/asr1/run.sh#L130 ?
Now we use `remove_longshortdata.sh` to remove long sentences.

**Comment by jnishi** [comment]
@sw005320 Thanks for useful comments. I fixed them. Please check again.
